### DISCUSSION TITLE
firefly-iii on k3s

### DISCUSSION THREAD

**Discussion by nebula2** [firefly-iii/firefly-iii#4778] [General]
Hey,
i am currently building a k3s cluster on my raspberry pi's. I also want to deploy firefly to this cluster. i saw that there is a k8s repo containing the yaml files. 

When I am done building this I wanted to write some articles in my blog about it. I plan to make a dedicated post for firefly-iii. Would that be okay? 

It is not the case that I have any audience but building this cluster grown to a little passion and that is why i wanted to share as much as I can.

Another question: If I adjust the yaml files in the k8s repo would it be okay if I make a PR for that? (just in case I can improve something)

**Comment by JC5** [comment]
Yes to both questions, feel free! Very cool 😎

**Comment by nebula2** [comment]
Little update:
today I managed backup and restore. 

I am still playing around with a few things and trying stuff out to see what changes have which impact.

but I have firefly running with everything needed including desaster recovery. as soon as I finished everything there will be an update :)

**Comment by nebula2** [comment]
sorry it has been a while >.<

i am so distracted with stuff that i just can not produce an outcome that is shareable.

however my cousin and me are both using the charts i created and the do their job.

some things that need to be worked on:
- use secrets for sensible information
- use stateful set for db instead of deployment with replicas strictly set to 1
- abstract backup/ restore logic
- seperate configmaps

and probably some other stuff xD

I would however like to create a branch containing my version as a reference for someone else to have something to get started with.

maybe someone with more time can finalize it. I would also work on the branch from time to time instead of maintaining my own version.

**Reply by JC5** [reply]
Tell me to which repositories to link (is Firefly III helper something I can add) and I can add them to the documentation. Would that work for you?

**Reply by nebula2** [reply]
i found the kubernetes repo yesterday.

for the documentation part: i can would also write docs as descriptive as i can. someone else should review them because my english is not the best :D


wanted to create a pr but i have some weird issue that i wanted to take a look at. seems like there is something wrong with my db.

**Reply by JC5** [reply]
Alright, let me know if I can help. Take your time, no rush!

**Reply by nebula2** [reply]
Found the issue. Forgot to set the environment variable DB_USERNAME leading to a funny error :D

```
root@firefly-firefly-iii-77dd76cdb6-hj4kl:/var/www/html/storage/logs# cat ff3-cli-2021-09-03.log
[2021-09-03 22:53:40] local.INFO: Could not create or generate installation ID. Do not continue.  
[2021-09-03 22:53:40] local.ERROR: Could not poll the database: SQLSTATE[08006] [7] FATAL:  role "www-data" does not exist (SQL: select "id", "name", "data" from "configuration" where "name" = is_demo_site and "configuration"."deleted_at" is null limit 1) {"exception":"[object] (FireflyIII\\Exceptions\\FireflyException(code: 0): Could not poll the database: SQLSTATE[08006] [7] FATAL:  role \"www-data\" does not exist (SQL: select \"id\", \"name\", \"data\" from \"configuration\" where \"name\" = is_demo_site and \"configuration\".\"deleted_at\" is null limit 1) at /var/www/html/app/Support/FireflyConfig.php:83)
[stacktrace]
#0 /var/www/html/app/Http/Controllers/Controller.php(54): FireflyIII\\Support\\FireflyConfig->get()
#1 /var/www/html/app/Http/Controllers/HomeController.php(55): FireflyIII\\Http\\Controllers\\Controller->__construct()
#2 [internal function]: FireflyIII\\Http\\Controllers\\HomeController->__construct()
#3 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(877): ReflectionClass->newInstanceArgs()
#4 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(714): Illuminate\\Container\\Container->build()
#5 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(841): Illuminate\\Container\\Container->resolve()
#6 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(652): Illuminate\\Foundation\\Application->resolve()
#7 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(826): Illuminate\\Container\\Container->make()
#8 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Route.php(268): Illuminate\\Foundation\\Application->make()
#9 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Route.php(1019): Illuminate\\Routing\\Route->getController()
#10 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Route.php(980): Illuminate\\Routing\\Route->controllerMiddleware()
#11 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(712): Illuminate\\Routing\\Route->gatherMiddleware()
#12 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(688): Illuminate\\Routing\\Router->gatherRouteMiddleware()
#13 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(672): Illuminate\\Routing\\Router->runRouteWithinStack()
#14 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(636): Illuminate\\Routing\\Router->runRoute()
#15 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(625): Illuminate\\Routing\\Router->dispatchToRoute()
#16 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(166): Illuminate\\Routing\\Router->dispatch()
#17 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\\Foundation\\Http\\Kernel->Illuminate\\Foundation\\Http\\{closure}()
#18 /var/www/html/app/Http/Middleware/InstallationId.php(54): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#19 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): FireflyIII\\Http\\Middleware\\InstallationId->handle()
#20 /var/www/html/vendor/fideloper/proxy/src/TrustProxies.php(57): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#21 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Fideloper\\Proxy\\TrustProxies->handle()
#22 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#23 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php(31): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#24 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\ConvertEmptyStringsToNull->handle()
#25 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#26 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php(40): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#27 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\TrimStrings->handle()
#28 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#29 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle()
#30 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php(86): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#31 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\PreventRequestsDuringMaintenance->handle()
#32 /var/www/html/app/Http/Middleware/SecureHeaders.php(51): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#33 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): FireflyIII\\Http\\Middleware\\SecureHeaders->handle()
#34 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#35 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(141): Illuminate\\Pipeline\\Pipeline->then()
#36 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(110): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter()
#37 /var/www/html/public/index.php(76): Illuminate\\Foundation\\Http\\Kernel->handle()
#38 {main}
"} 
```

**Comment by nebula2** [comment]
Here's what I have so far:
[4778-k3s-example](https://github.com/nebula2/kubernetes/tree/4778-k3s-example)

EDIT:
[PR](https://github.com/firefly-iii/kubernetes/pull/7)

**Reply by JC5** [reply]
Nice!

**Comment by morremeyer** [comment]
I'm picking up the work on this right now. I'll take care of the TODOs @nebula2 noted and add some configurability for common scenarios.

@JC5, are you more a fan of more small PRs that gradually introduce changes or should I rather open less PRs? Changes will of course be grouped into meaninfully separated commits.

**Reply by JC5** [reply]
If you can do a meaningful group that would be great! That way changes are also easy to review for others 👍

**Reply by morremeyer** [reply]
Sounds good. I'll start with some changes that are necessary to get it running in different environments in one PR and then group further commits so that each meaningful change is one PR.

I'm always open for feedback, so if I get too rapid-fire on my PRs or else, ping me anytime.

**Reply by JC5** [reply]
Will do, thanks!

**Comment by morremeyer** [comment]
We've come quite far here already, I thought I'd write down some of the next steps I'm thinking about to gather feedback.

My ideas are:

- Use https://github.com/helm/chart-releaser-action to release the charts to GitHub pages as helm repository (this will need some work on chart-releaser as we have dependencies on charts in the same repo, see [this action run](https://github.com/morremeyer/kubernetes/runs/5139505528?check_suite_focus=true) for a fail with the current version. If upstream is not interested in a feature to support this, we need to think of something else
- Adding https://github.com/apps/renovate to automatically open PRs for updates of images in the values.yaml files, additional chart dependencies and updates for e.g. github actions dependencies and helm versions for chart-releaser actions. You can check out [my charts repository renovate.json](https://github.com/morremeyer/charts/blob/2ff091e60e97fbc89e0e9c4aa0c53c68be0de828/renovate.json) and the [chart linting and releasing configuration](https://github.com/morremeyer/charts/blob/2ff091e60e97fbc89e0e9c4aa0c53c68be0de828/.github/workflows/lint-and-release.yml) I use as well as some of the commit history in my charts repo for more context.
- Adding a workflow that fails when files in a chart have been changed without bumping the chart version so that we don't accidentally introduce changes to users

Feedback, especially yours @JC5, is very welcome here.

**Reply by JC5** [reply]
This looks like an excellent excuse for me to learn k8n. 😁

**Reply by morremeyer** [reply]
I'll start implementing the things I listed one after the other, as I've done already with the chart updates. I'll try to write comprehensive PR descriptions so that things are very clear.

Btw, can I ask why you call it k8n? It's usually called k8s, I also noticed this is the kubernetes repo description.

**Reply by JC5** [reply]
Oh just my little mistake, feel free to correct me. I keep saying k8n but i have no idea why 😂

**Reply by morremeyer** [reply]
@tyzbit Pulling you in here as you seem to be using the `firefly-db` chart:

Do you have an opinion on deprecating and removing the integrated `firefly-db` chart in favor of bitnami/postgresql with added documentation for backups?

**Reply by tyzbit** [reply]
Our needs can be met with `bitnami/postgresql` and if we use it users are more likely to already have some familiarity with that chart in case they need to adjust settings. I support deprecating the existing db chart and recommending users just use `bitnami/postgresql`.

Backing up PVCs is something the operator of any k8s cluster should be aware of and look into, and in my opinion is outside the scope of what should be included in the chart. I think simply providing documentation and links on backing up the database (or really any persistent data) is appropriate. Plus, any backup solution we did provide would inevitably be inadequate for some configurations, or would outright fail in others. Best to educate, rather than facilitate, backing up important data.
